### PR TITLE
[smb] Move firewall_checker to built-in function & Rewrite RDP module #66 & bugs fix

### PR DESCRIPTION
**Pull request by XiaoliChan**
Changes:
- Move `firewall_checker` into the built-in function as `dcom_FirewallChecker` 
- SMB/AT/MMC command execution exceptions review.
- Firewall check in MMCExec.
- RDP module has been rewritten #66

### CODE REVIEW

**Comment by XiaoliChan** [comment]
I think moving the `firewall_checker` into the built-in function is the best way for somebody who wants to write smb module with wmi.

**Comment by XiaoliChan** [comment]
@bongobongoland can you give a test about the new RDP module?

**Comment by XiaoliChan** [comment]
Target system: server 2019
image

For old system: server 2003
image

**Comment by bongobongoland** [comment]
I'm on it

**Comment by XiaoliChan** [comment]
> What does OLD=true mean?

For the old system like server 2003 (under NT6)

**Comment by XiaoliChan** [comment]
> 

Nice catch!

**Comment by XiaoliChan** [comment]
Can you test with disabled firewall?

**Comment by bongobongoland** [comment]
Works fine now. BUT I had to "netsh advfirewall set allprofiles state off"
I think the module should either disable the firewall (not pretty) or "netsh firewall set service type = remotedesktop mode = enable" (prettier) or something like that, as I suggested earlier.

**Comment by XiaoliChan** [comment]
> Works fine now. BUT I had to "netsh advfirewall set allprofiles state off" I think the module should either disable the firewall (not pretty) or "netsh firewall set service type = remotedesktop mode = enable" (prettier) or something like that, as I suggested earlier.

Hmm, I still try to figure out how to enable RDP in smb without touching CMD

**Comment by bongobongoland** [comment]
So...  running "netsh firewall set service type = remotedesktop mode = enable" is not enough. obviously. 
It seems that the firewall has to be disabled completely, only then wmi works fine.

**Comment by XiaoliChan** [comment]
> So... running "netsh firewall set service type = remotedesktop mode = enable" is not enough. obviously. It seems that the firewall has to be disabled completely, only then wmi works fine.

I have read a post about using the registry to control firewall rules, but it didn't work......

**Comment by XiaoliChan** [comment]
@bongobongoland BTW, can you test again with this PR with firewall enabled.

**Comment by bongobongoland** [comment]
I'm on it.

**Comment by XiaoliChan** [comment]
> I'm on it.

any feedback？

**Comment by bongobongoland** [comment]
First find. Well. Not really :) Some older machines need the following:

```
reg add "HKLM\SYSTEM\CurrentControlSet\Control\Terminal Server" /v IgnoreRegUserConfigErrors /t REG_DWORD /d 1 /f
```

But I guess that is not relevant to your code :)

**Comment by bongobongoland** [comment]
ACTION=status would be good.
Spits out, say, the following: TermService is already running on port 3395

**Comment by bongobongoland** [comment]
I'm still testing it.

> > I'm on it.
> 
> any feedback？

**Comment by XiaoliChan** [comment]
> 

Ignore certificates? you can do this in your xfreerdp with `/cert-ignore /tls-seclevel:0 /timeout:80000`

**Comment by XiaoliChan** [comment]
> ACTION=status would be good. Spits out, say, the following: TermService is already running on port 3395

Hmmm, not going to do this, this needs to do another query stuff...

**Comment by XiaoliChan** [comment]
> First find. Well. Not really :) Some older machines need the following:
> 
> ```
> reg add "HKLM\SYSTEM\CurrentControlSet\Control\Terminal Server" /v IgnoreRegUserConfigErrors /t REG_DWORD /d 1 /f
> ```
> 
> But I guess that is not relevant to your code :)

Any reference to this registry?

**Comment by bongobongoland** [comment]
> > 
> 
> Ignore certificates? you can do this in your xfreerdp with `/cert-ignore /tls-seclevel:0 /timeout:80000`

no, IgnoreRegUserConfigErrors is about https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/troubleshoot-rdp-access-denied

**Comment by XiaoliChan** [comment]
> no, IgnoreRegUserConfigErrors is about https://learn.microsoft.com/en-us/troubleshoot/azure/virtual-machines/troubleshoot-rdp-access-denied

Well, I didn't see that case, can you pint a screenshot of this error?

**Comment by bongobongoland** [comment]
I agree.
So for I'm not seeing any errors with the module.

**Comment by XiaoliChan** [comment]
@NeffIsBack @mpgn Can do review now!

File: cme/protocols/smb.py [hunk 1]
```diff
@@ -1207,45 +1211,55 @@ def loggedon_users(self):
     def pass_pol(self):
         return PassPolDump(self).dump()
 
+    @requires_admin
     def wmi(self, wmi_query=None, namespace=None):
```
Comment (mpgn) [inline cme/protocols/smb.py:1215]: i think we can remove this function if we go with a new proto wmi

Comment (XiaoliChan) [reply cme/protocols/smb.py:1215]: > i think we can remove this function if we go with a new proto wmi

Sure!

Comment (XiaoliChan) [reply cme/protocols/smb.py:1215]: Do you mean remove the whole `wmi` query function in smb?

Comment (XiaoliChan) [reply cme/protocols/smb.py:1215]: I will make a new PR, if the `WMI` protocol is merged, to many things in this PR.

**Review by mpgn — CHANGES_REQUESTED** [review]
stacktrace with option --wmi

File: cme/modules/rdp.py [hunk 2]
```diff
@@ -83,24 +167,170 @@ def rdp_disable(self, context, smbconnection):
             ans = rrp.hBaseRegOpenKey(
                 remoteOps._RemoteOperations__rrp,
                 regHandle,
-                "SYSTEM\\CurrentControlSet\\Control\\Terminal Server",
+                "System\\CurrentControlSet\\Control\\Lsa",
```
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py:170]: Yes, this is for restricted admin

**Comment by mpgn** [comment]
I feel like this dns related (aka because no dns) but it should not popup a stacktrace like this

**Comment by XiaoliChan** [comment]
> Not better

Weird, Looks like making the Dcom connection failed on your target.

What happen if you try to execute command via wmiexec?

**Comment by XiaoliChan** [comment]
Ohhhh, I know where is the issues XD

**Comment by mpgn** [comment]
But what i'm looking for is a clean error instead of full stracktrace

File: cme/protocols/smb.py [hunk 3]
```diff
@@ -56,7 +56,10 @@
 from dploot.lib.smb import DPLootSMBConnection
 
 from pywerview.cli.helpers import *
-from pywerview.requester import RPCRequester
+
+from impacket.dcerpc.v5.dtypes import NULL
```
Comment (mpgn) [inline cme/protocols/smb.py:60]: place them with all the impacket line just before :)

**Review by mpgn — CHANGES_REQUESTED** [review]

**Comment by XiaoliChan** [comment]
Please test again, I forgot change the `self.conn.getRemoteHost() to self.host`

**Comment by XiaoliChan** [comment]
> Will it be compatible with Kerberos ? Maybe `self.host if not self.kerberos else self.hostname + "." + self.domain` is better

Awesome !

**Comment by XiaoliChan** [comment]
> Will it be compatible with Kerberos ? Maybe `self.host if not self.kerberos else self.hostname + "." + self.domain` is better

Done!

**Comment by mpgn** [comment]
image

🎉

**Comment by mpgn** [comment]
image


🎉

**Comment by XiaoliChan** [comment]
> image
> 
> 🎉

Can you try "-o METHOD=smb ACTION=enable", make sure I not break the `ncacn_np` way

**Review by mpgn — DISMISSED** [review]

**Comment by mpgn** [comment]
image

🎉

**Comment by XiaoliChan** [comment]
> image
> 
> 🎉

Awesome!

**Comment by XiaoliChan** [comment]
@NeffIsBack Hello Neff, please do a review about this PR (think there were too many messages in your box that made you forget this 😁)

**Comment by XiaoliChan** [comment]
> @XiaoliChan i already left comments :D
> 
> image

Hmm, weird, I can't see that

**Comment by NeffIsBack** [comment]
> Hmm, weird, I can't see that

Ah that might be my fault, sec

**Comment by XiaoliChan** [comment]
> image

The SMB, means (ncacn_np), the WMI means (ncacn_tcp_ip), also can mean "executing using smb"

File: cme/modules/rdp.py [hunk 4]
```diff
@@ -21,26 +28,97 @@ def __init__(self, context=None, module_options=None):
 
     def options(self, context, module_options):
         """
-        ACTION       Enable/Disable RDP (choices: enable, disable)
+        ACTION       Enable/Disable RDP (choices: enable, disable, enable-ram, disable-ram)
+        METHOD       wmi(ncacn_ip_tcp)/smb(ncacn_np) (choices: wmi, smb, default is wmi)
```
Comment (NeffIsBack) [inline cme/modules/rdp.py:32]: The Option comments for "OLD" and "DCOM-TIMEOUT" are missing

**Comment by XiaoliChan** [comment]
> Ah that might be my fault, sec

Not sure what happen
image

File: cme/modules/rdp.py [hunk 5]
```diff
@@ -21,26 +28,97 @@ def __init__(self, context=None, module_options=None):
...
+            try:
+                smb_rdp = rdp_SMB(context, connection)
+                if "ram" in self.action:
+                    smb_rdp.rdp_RAMWrapper(self.action)
+                else:
+                    smb_rdp.rdp_Wrapper(self.action)
+            except Exception as e:
+                context.log.fail(f"Enable RDP via smb error: {str(e)}")
+        elif self.method == "wmi":
+            context.log.info("Doing action with WMI(ncacn_ip_tcp)")
```
Comment (NeffIsBack) [inline cme/modules/rdp.py:83]: Same for this, i think that should be more concret

**Review by NeffIsBack — CHANGES_REQUESTED** [review]
Rest looks good to me

File: cme/modules/rdp.py [hunk 6]
```diff
@@ -21,26 +28,97 @@ def __init__(self, context=None, module_options=None):
...
+            self.oldSystem = True
 
     def on_admin_login(self, context, connection):
-        if self.action == "enable":
-            self.rdp_enable(context, connection.conn)
-        elif self.action == "disable":
-            self.rdp_disable(context, connection.conn)
+        # Preparation for wmi protocol
+        if self.method == "smb":
+            context.log.info("Doing action with SMB(ncacn_np)")
```
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py:73]: Yes, should I change it to "Executing over SMB"?

**Comment by NeffIsBack** [comment]
> The SMB, means (ncacn_np), the WMI means (ncacn_tcp_ip), also can mean "executing using smb"

Ah okay, maybe something like "Using SMB with ncacn_np" or "Executing commands using WMI (ncacn_tcp_ip)"?

**Comment by XiaoliChan** [comment]
> Ah okay, maybe something like "Using SMB with ncacn_np" or "Executing commands using WMI (ncacn_tcp_ip)"?

Yes

File: cme/modules/rdp.py (hunk 5 above)
Comment (XiaoliChan) [reply cme/modules/rdp.py]: Done!

File: cme/modules/rdp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (XiaoliChan) [reply cme/modules/rdp.py]: Fixed!

**Comment by NeffIsBack** [comment]
Ah would be nice if you would add the default value for dcom-timeout

**Comment by XiaoliChan** [comment]
> Ah would be nice if you would add the default value for dcom-timeout (To the module option comment)

Already add in line 59 https://github.com/mpgn/CrackMapExec/pull/114/files#diff-b5c2077d8eda2116d690da1dba11b2bd82d473b5d6bcf88d421535e8bb0217f7R59

**Comment by NeffIsBack** [comment]
> > Ah would be nice if you would add the default value for dcom-timeout (To the module option comment)
> 
> Already add in line 59 https://github.com/mpgn/CrackMapExec/pull/114/files#diff-b5c2077d8eda2116d690da1dba11b2bd82d473b5d6bcf88d421535e8bb0217f7R59

sorry, i mean't the module option comment. Saw your default value in the code and i think it would be good for the user to know that it already has a default value (so the user does not need to specify it)

**Comment by XiaoliChan** [comment]
> sorry, i mean't the module option comment. Saw your default value in the code and i think it would be good for the user to know that it already has a default value (so the user does not need to specify it)

Ha, I get it!

**Review by NeffIsBack — APPROVED** [review]
LGTM 👍

**Comment by XiaoliChan** [comment]
@NeffIsBack Fixed!

**Review by mpgn — APPROVED** [review]

**Comment by bongobongoland** [comment]
When running -M rdp -o ACTION=enable

**Comment by XiaoliChan** [comment]
> 
> When running -M rdp -o ACTION=enable

Looks like something is wrong with DCOM, can you run wmiexec.py with --debug

**Comment by bongobongoland** [comment]
like this?

**Comment by XiaoliChan** [comment]
> like this? 

No just impacket's wmiexec.py
Please open a new issues ticket.

**Comment by bongobongoland** [comment]
Impacket's wmiexec works fine.

**Comment by XiaoliChan** [comment]
> Impacket's wmiexec works fine.

Please open a new issues ticket and show me the details.

**Pull request merged**